### ISSUE TITLE
Job failure back-off delay and limit are not configurable

### ISSUE THREAD

**Issue by sathyanarays** [kubernetes/kubernetes#114651]
### What happened?

The job failure back-off parameters are computed based on the [hardcoded constants in job controller](https://github.com/kubernetes/kubernetes/blob/a6d66d15f691cb6473f1361caed08e5bd7eae68b/pkg/controller/job/job_controller.go#L75-L77). There is no way to override these back-off parameters at Job level

### What did you expect to happen?

The user should have the ability to provide back-off parameters as part of the Job Spec

### How can we reproduce it (as minimally and precisely as possible)?

NA

### Kubernetes version

NA

### Cloud provider

NA

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by sathyanarays** [comment]
related to #114391 and #114650

**Comment by alculquicondor** [comment]
This could be good additions to the `PodFailurePolicy` struct https://github.com/kubernetes/kubernetes/blob/2bb77a13b1a709e0208c6ba9ba5323a5e54f79d6/staging/src/k8s.io/api/batch/v1/types.go#L199

**Comment by alculquicondor** [comment]
@zeusbee note that this change requires a KEP

**Comment by xadhix** [comment]
Hi @alculquicondor @zeusbee I created a KEP for this. Could you please take a look at it?

**Comment by alculquicondor** [comment]
@sathyanarays @xadhix, what configuration of backoff do you expect to use?

I wonder if it's feasible and good enough for you to just reduce the backoff that we currently have, or make it purely exponential: 1s, 2s, 4s, 8s, etc., as suggested @mimowo in another thread.

Since the backoff wasn't properly working until recently, I don't expect users to currently rely on the existing delays.

**Comment by atiratree** [comment]
Is there a precedence for this? Do users really care about setting a backoff for their Jobs? Wouldn't it make more sense to add such a low-level option to kube-controller-manager instead? And leave the responsibility of deciding backoffs to the cluster/platform admin?

Nevertheless, allowing to pass the value for the tests seems reasonable.

**Comment by alculquicondor** [comment]
I agree that this is not something that should be exposed in the Job API. I suggested earlier that we just reduce the hardcoded backoff. But a configuration, along with a change of defaults could be good too.

**Comment by tenzen-y** [comment]
My primary question is, do we want to use these parameters via Job API? The config (https://github.com/kubernetes/kubernetes/blob/dc481fef903ac4d1b486fc0aa39795ad1f3d3213/pkg/controller/job/config/types.go#L20) or kube-controller-manager command args might be enough place.

But I'm open to this discussion.

**Comment by atiratree** [comment]
> I suggested earlier that we just reduce the hardcoded backoff. 

Yeah, that could be enough as well. Although I am not sure what was the reasoning behind the 10s originally. 

> The config  or kube-controller-manager command arg

Usually the args go hand in hand with the config.

**Comment by mimowo** [comment]
I suppose users don't need per-job control, I was looking for a solution to reduce runtime of e2e tests, and piggybacked on this old issue :).

For the e2e runtime use-case reducing the default base in code to 1s would be great. Maybe @soltysh knows what was the reason for 10s.

Assuming we go with reducing the default base, the question remains if we need some safety mechanism for users who would rely on 10s (TBH I doubt anyone relied on this as pointed out by Aldo above), and if so, what do we choose: a feature gate or kube-controller-manager option. Does adding a new kube-controller-manager option require a KEP from the process perspective?

**Comment by mimowo** [comment]
cc @kaisoz who is interested in implementing it once we decide on the approach

**Comment by soltysh** [comment]
I'm inline with Aldo and Filip on this matter, we should NOT expose such a configuration to users. Based on the above it looks like the only use-case are currently our e2e-s, which also seem like rather weak use-case. The fact that tests would rely on specific controller settings would make it complicated to run these for some users validating functionality in their clusters. 

If we care about restructuring these tests I'd probably consider moving some to integration or look into other options rather than changing cluster defaults.

**Comment by soltysh** [comment]
I was talking with Michal on slack about it, and to make it explicitly clear, I've meant also cluster settings and defaults when stating the above. My main worry about changing defaults from 10s to 1s is that we can increase the load 3x (with 1s, 2s, 4s delays if we go with 1s initial). That's a significant load that I'm not sure I see sufficient proof we need at this point in time. Similar worries apply to cluster setting, especially that the only use-case I hear for now is tests.
